File: test/datatest.cpp
#include "gtest/gtest.h"

#include "entry.h"
#include "data.h"

/* Single publish. Alice's timeline request should return all Alice's posts */
TEST(Data, Test1) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Alice", "I love the weather today.", t_time);
    std::vector<Entry> t_timeline = t_data.timeline("Alice");

    EXPECT_EQ(t_timeline.size(), 1u);
}

/* Single publish. Alice timeline request should return posts owned by Alice  */
TEST(Data, Test2) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Alice", "I love the weather today.", t_time);
    std::vector<Entry> t_timeline = t_data.timeline("Alice");

    EXPECT_EQ(t_timeline[0].user, "Alice");
}

/* Single publish. Alice timeline request should return Alice's post content */
TEST(Data, Test3) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Alice", "I love the weather today.", t_time);
    std::vector<Entry> t_timeline = t_data.timeline("Alice");

    EXPECT_EQ(t_timeline[0].message, "I love the weather today.");
}

/* Single publish. Alice timeline request should return the time of Alice's post */
TEST(Data, Test4) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Alice", "I love the weather today.", t_time);
    std::vector<Entry> t_timeline = t_data.timeline("Alice");

    EXPECT_EQ(t_timeline[0].timestamp, t_time);
}

/* Multiple publish.  Alice's timeline request should return all Alice's posts */
TEST(Data, Test5) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Alice", "I love the weather today.", t_time);
    t_data.publish("Alice", "I love the weather again.", t_time);
    std::vector<Entry> t_timeline = t_data.timeline("Alice");

    EXPECT_EQ(t_timeline.size(), 2u);
}

/* Multiple publish. Alice timeline request should return posts owned by Alice */
TEST(Data, Test6) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Alice", "I love the weather today.", t_time);
    t_data.publish("Alice", "I love the weather again.", t_time);
    std::vector<Entry> t_timeline = t_data.timeline("Alice");

    EXPECT_EQ(t_timeline[0].user, "Alice");
    EXPECT_EQ(t_timeline[1].user, "Alice");
}

/* Multiple publish. Alice timeline request should return correct post content */
TEST(Data, Test7) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Alice", "I love the weather today.", t_time);
    t_data.publish("Alice", "I love the weather again.", t_time);
    std::vector<Entry> t_timeline = t_data.timeline("Alice");

    EXPECT_EQ(t_timeline[1].message, "I love the weather again.");
}

/* Single publish. Alice timeline request should not return any of Bob's posts */
TEST(Data, Test8) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Bob", "Hello.", t_time);
    std::vector<Entry> t_timeline = t_data.timeline("Alice");

    EXPECT_EQ(t_timeline.size(), 0u);
}

/* Single publish. Bob timeline request should return all of Bob's posts */
TEST(Data, Test9) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Bob", "Hello.", t_time);
    std::vector<Entry> t_timeline = t_data.timeline("Bob");

    EXPECT_EQ(t_timeline.size(), 1u);
}

/* Single publish. Bob timeline request should return posts owned by Bob */
TEST(Data, Test10) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Bob", "Hello.", t_time);
    std::vector<Entry> t_timeline = t_data.timeline("Bob");

    EXPECT_EQ(t_timeline[0].user, "Bob");
}

/* Multiple user publish. Bob timeline request should return posts owned by Bob */
TEST(Data, Test11) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Alice", "I love the weather today.", t_time);
    t_data.publish("Bob", "Hello.", t_time);
    t_data.publish("Alice", "I love the weather again.", t_time);

    std::vector<Entry> t_timeline = t_data.timeline("Bob");

    EXPECT_EQ(t_timeline.size(), 1u);
}

/* Multiple user publish. Alice timeline request should return posts owned by Alice */
TEST(Data, Test12) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Alice", "I love the weather today.", t_time);
    t_data.publish("Bob", "Hello.", t_time);
    t_data.publish("Alice", "I love the weather again.", t_time);

    std::vector<Entry> t_timeline = t_data.timeline("Alice");

    EXPECT_EQ(t_timeline.size(), 2u);
}

/* Single publish. Alice timeline request should ordered from youngest to oldest */
TEST(Data, Test13) {
    Data t_data;
    std::time_t t_time = std::time(nullptr);

    t_data.publish("Alice", "I love the weather today.", t_time-2);
    t_data.publish("Alice", "I love the weather again.", t_time);

    std::vector<Entry> t_timeline = t_data.timeline("Alice");

    EXPECT_EQ(t_timeline[0].message, "I love the weather again.");
}

File: java/codechallenge/src/test/java/codechallenge/impl/DataTest.java
package codechallenge.impl;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.util.Date;
import java.util.List;

import org.junit.Before;
import org.junit.Test;

import codechallenge.interfaces.DataInterface;

public class DataTest {

	private DataInterface t_data;
	
	@Before
	public void setup() {
		t_data = new Data();
	}
	
	@Test
	public void test01() {
		Date t_time = new Date(0);
		
		t_data.publish("Alice", "I love the weather today.", t_time);
	    List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertTrue("Timeline should not be null", t_timeline != null);
	}
	
	@Test
	public void test02() {
		Date t_time = new Date(0);
		
		t_data.publish("Alice", "I love the weather today.", t_time);
	    List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertEquals("Incorrect timeline size", 1, t_timeline.size());
	}
	
	@Test
	public void test03() {
		Date t_time = new Date(0);
		
		t_data.publish("Alice", "I love the weather today.", t_time);
	    List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertEquals("Incorrect timeline user", "Alice", t_timeline.get(0).getUser());
	}
	
	@Test
	public void test04() {
		Date t_time = new Date(0);
		
		t_data.publish("Alice", "I love the weather today.", t_time);
	    List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertEquals("Incorrect timeline message", "I love the weather today.", t_timeline.get(0).getMessage());
	}
	
	@Test
	public void test05() {
		Date t_time = new Date(0);
		
		t_data.publish("Alice", "I love the weather today.", t_time);
	    List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertEquals("Incorrect timeline timestamp", t_time, t_timeline.get(0).getTimestamp());
	}
	
	@Test
	public void test06() {
		Date t_time = new Date(0);
		
		t_data.publish("Alice", "I love the weather today.", t_time);
	    t_data.publish("Alice", "I love the weather again.", t_time);
	    List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertEquals("Incorrect timeline size", 2, t_timeline.size());
	}
	
	@Test
	public void test07() {
		Date t_time = new Date(0);
		
		t_data.publish("Alice", "I love the weather today.", t_time);
	    t_data.publish("Alice", "I love the weather again.", t_time);
	    List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertEquals("Incorrect timeline user 0", "Alice", t_timeline.get(0).getUser());
	    assertEquals("Incorrect timeline user 1", "Alice", t_timeline.get(1).getUser());
	}
	
	@Test
	public void test08() {
		Date t_time = new Date(0);
		
		t_data.publish("Alice", "I love the weather today.", t_time);
	    t_data.publish("Alice", "I love the weather again.", t_time);
	    List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertEquals("Incorrect timeline message", "I love the weather again.", t_timeline.get(1).getMessage());
	}
	
	@Test
	public void test09() {
		Date t_time = new Date(0);
		
		t_data.publish("Bob", "Hello.", t_time);
		List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertEquals("Incorrect timeline size", 0, t_timeline.size());
	}
	
	@Test
	public void test10() {
		Date t_time = new Date(0);
		
		t_data.publish("Bob", "Hello.", t_time);
		List<Entry> t_timeline = t_data.timeline("Bob");
		
	    assertEquals("Incorrect timeline size", 1, t_timeline.size());
	}
	
	@Test
	public void test11() {
		Date t_time = new Date(0);
		
		t_data.publish("Bob", "Hello.", t_time);
		List<Entry> t_timeline = t_data.timeline("Bob");
		
	    assertEquals("Incorrect timeline user", "Bob", t_timeline.get(0).getUser());
	}
	
	@Test
	public void test12() {
		Date t_time = new Date(0);
		
		t_data.publish("Alice", "I love the weather today.", t_time);
	    t_data.publish("Bob", "Hello.", t_time);
	    t_data.publish("Alice", "I love the weather again.", t_time);
	    
	    List<Entry> t_timeline = t_data.timeline("Bob");
		
	    assertEquals("Incorrect timeline size", 1, t_timeline.size());
	}
	
	@Test
	public void test13() {
		Date t_time = new Date(0);
		
		t_data.publish("Alice", "I love the weather today.", t_time);
	    t_data.publish("Bob", "Hello.", t_time);
	    t_data.publish("Alice", "I love the weather again.", t_time);
	    
	    List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertEquals("Incorrect timeline size", 2, t_timeline.size());
	}
	
	@Test
	public void test14() {
		Date t_time1 = new Date(0);
		Date t_time2 = new Date(2000);
		
		t_data.publish("Alice", "I love the weather today.", t_time1);
	    t_data.publish("Alice", "I love the weather again.", t_time2);

	    List<Entry> t_timeline = t_data.timeline("Alice");
		
	    assertEquals("Timelime messages out of order", "I love the weather again.", t_timeline.get(0).getMessage());
	}
}
